### DISCUSSION TITLE
Typing support for `relationship` function

### DISCUSSION THREAD

**Discussion by michaelbukachi** [sqlalchemy/sqlalchemy#8732] [Usage Questions] [answered]
Hello.
Does SQLAlchemy 2.0 have typing support for the `relationship` function? 
I don't seem to understand this snippet from the documentation.
![screenshot_20221028173657](https://user-images.githubusercontent.com/10145850/198674376-4d1f2560-8755-4249-8a56-847713ac617d.png)

When running mypy, I get the following error:
```bash
 Incompatible types in assignment (expression has type "RelationshipProperty[<nothing>]", variable has type "List[ModelA]"
```

**Comment by zzzeek** [comment, marked as answer]
> Hello. Does SQLAlchemy 2.0 have typing support for the `relationship` function? I don't seem to understand this snippet from the documentation. ![screenshot_20221028173657](https://user-images.githubusercontent.com/10145850/198674376-4d1f2560-8755-4249-8a56-847713ac617d.png)

the warning refers to the feature of Declarative where an attribute may be mapped at runtime by joining information from both sides of an annotated expression into one, such as 

```py

intpk = Annotated[int, mapped_column(primary_key=True)]

class MyClass(Base):
    __tablename__ = 'my_table'
    # ...

    # will automatically map "id" to an integer column with the primary key
    # parameter set to True
    id: Mapped[intpk]
```

that is, it refers strictly to runtime Python behavior and has nothing to do with Mypy.   I will rephrase the paragraph to hopefully eliminate ambiguity as follows:

```
.. note:: The feature of :func:`_orm.mapped_column` just described, where
   a fully constructed set of column arguments may be indicated using
   :pep:`593` ``Annotated`` objects that contain a "template"
   :func:`_orm.mapped_column` object to be copied into the attribute, is
   currently not implemented for other ORM constructs such as
   :func:`_orm.relationship` and :func:`_orm.composite`.   While this functionality
   is in theory possible, for the moment attempting to use ``Annotated``
   to indicate further arguments for :func:`_orm.relationship` and similar
   will raise a ``NotImplementedError`` exception at runtime, but
   may be implemented in future releases.
```



> 
> When running mypy, I get the following error:
> 
> ```shell
>  Incompatible types in assignment (expression has type "RelationshipProperty[<nothing>]", variable has type "List[ModelA]"
> ```

so that's a typing error and I would need to see the code that is being typed as such, and it might be that you need to make sure you use the ``Mapped[]`` annotation.  at the same time you want to make sure you aren't using the SQLAlchemy mypy plugin anymore if this is 2.0 style code and also that you don't have any sqlalchemy-stubs packages installed.  All of the examples in SQLAlchemy's documentation should pass mypy without errors, including the [quickstart](https://docs.sqlalchemy.org/en/20/orm/quickstart.html#declare-models) model which you can paste into a file and test that mypy produces no errors.

**Reply by michaelbukachi** [reply]
When I copy the snippet and run mypy against it I get the following error:
![screenshot_20221028200103](https://user-images.githubusercontent.com/10145850/198692314-e63d5ae6-cd3d-413a-b1ba-6fea9d9d9d0a.png)

**Reply by zzzeek** [reply]
right then you are definitely using stubs such as [sqlalchemy2-stubs](https://github.com/sqlalchemy/sqlalchemy2-stubs).   Make sure mypy is set up to not have access to these .

**Reply by michaelbukachi** [reply]
Ah. I see. It was still in my `venv` directory. Had to explicitly uninstall it.
